runtime/src/accounts_db.rs

```diff
@@ -185,27 +188,57 @@ impl Versioned for (u64, AccountInfo) {
     }
 }
 
+#[derive(Debug)]
 pub enum LoadedAccountAccessor<'a> {
+    // StoredAccountMeta can't be held directly here due to its lifetime dependency to
+    // AccountStorageEntry
     Stored(Option<(Arc<AccountStorageEntry>, usize)>),
-    Cached((&'a AccountsCache, Slot, &'a Pubkey)),
+    // None value in Cached variant means the cache was flushed
+    Cached(Option<(Pubkey, Cow<'a, CachedAccount>)>),
 }
 
 impl<'a> LoadedAccountAccessor<'a> {
-    fn get_loaded_account(&self) -> Option<LoadedAccount> {
+    fn check_and_get_loaded_account(&mut self) -> LoadedAccount {
         match self {
-            LoadedAccountAccessor::Stored(storage_entry) => {
-                // May not be present if slot was cleaned up in between
-                storage_entry.as_ref().and_then(|(storage_entry, offset)| {
-                    storage_entry
-                        .get_stored_account_meta(*offset)
-                        .map(LoadedAccount::Stored)
-                })
+            LoadedAccountAccessor::Cached(None) | LoadedAccountAccessor::Stored(None) => {
+                panic!("Should have already been taken care of when creating this LoadedAccountAccessor");
             }
-            LoadedAccountAccessor::Cached((cache, slot, pubkey)) => {
-                // May not be present if slot was cleaned up in between
-                cache.load(*slot, pubkey).map(|cached_account| {
-                    LoadedAccount::Cached((**pubkey, Cow::Owned(cached_account)))
-                })
+            LoadedAccountAccessor::Cached(Some(_cached_account)) => {
+                // Cached(Some(x)) variant always produces `Some` for get_loaded_account() since
+                // it just returns the inner `x` without additional fetches
+                self.get_loaded_account()
+                    .expect("Cache flushed should be handled before trying to fetch account")
```

Comment:
Previously, this `.expect()` existed in `get_loaded_account`. But, for the rpc case, this `panic!`s. So, I moved here. Also, I enhanced a test https://github.com/solana-labs/solana/pull/15454/files?diff=split&w=1#r606629658

Reply:
oh hmm, for the RPC panic:

1) is it happening from a scan like `scan_accounts()`/`unchecked_scan_accounts()`. Hopefully these scans shouldn't encounter this panic because they hold the index lock: https://github.com/solana-labs/solana/blob/master/runtime/src/accounts_index.rs#L507-L515. If this is panicking worried this means that RPC is getting inconsistent state.

2)And double checking here. For point lookups/fetch of a specific account through `do_load()`, even if we hit the max retryable iterations, then this should just return: https://github.com/solana-labs/solana/blob/3e0377b45867a2579710632fe00099fb13c3322d/runtime/src/accounts_db.rs#L2581 before calling `check_and_get_loaded_account()`

Reply:
@carllin 

> for the RPC panic:

good question. From this discussion, I think we can revert to your original change and even simplify more.

> 1. is it happening from a scan like `scan_accounts()`/`unchecked_scan_accounts()`. Hopefully these scans shouldn't encounter this panic because they hold the index lock: 

This is correct. The only concern I had was `purging` ( https://github.com/solana-labs/solana/pull/15454/files?diff=split&w=1#r606978231 ), which does irregular update order of `cache/store remove => index` update unlike other race-condition triggers. In conclusion, I think we can promote the purging race-condition detection a hard error (https://github.com/solana-labs/solana/pull/15454/files?diff=split&w=1#r606630573).

let me explain in pieces. :)

My original motivation of this `.expect()` move was that I thought we need to consider the purging case because of the existence of the purging race-condition detection: If `do_load(with_fixed_root = false)` from rpc must be handled for the purging, we likewise also must handle the purging case for the `scan_accounts()` and its families.

if that's true (I think it's not now), holding the index lock isn't enough because the purging code firstly removes caches/stores from the store:  https://github.com/solana-labs/solana/pull/15454/files?diff=split&w=1#r606978231

But then, I now think there should be no possibilities of purging race to begin with: All of code path for both `do_load` and `scan` actually ensures Bank's liveliness because they're called via as a method of `Bank` from external subsystems (rpc, replay&banking). So, combined with `Bank::parent` fields, all of ancestors should be alive then none of them should be purged/pruned. 


![image](https://user-images.githubusercontent.com/117807/113565857-aa103c80-9646-11eb-929a-440225d4d8f5.png)

![image](https://user-images.githubusercontent.com/117807/113566459-afba5200-9647-11eb-8602-2d8fa3ab8f5d.png)



With these things are done, am I right? Or, did you add the original purging detection code in `do_load` ( https://github.com/solana-labs/solana/commit/2885d989a4d6a93fcce0adca7f6210709550184e#r49109545 ) with some concrete concerns in mind at the time? 


> 2) And double checking here. For point lookups/fetch of a specific account through `do_load()`, even if we hit the max retryable iterations, then this should just return:

Yeah, this should be safe. So, in this regard too, we can restore the `.expect` in `get_loaded_account()`.

Reply:
Also, It means the purge race-condition test is meaningless because it's doing which can't happen in the wild. xD

Reply:
All these nice links aren't working for me 😢 

> But then, I now think there should be no possibilities of purging race to begin with: All of code path for both `do_load` and `scan` actually ensures Bank's liveliness because they're called via as a method of `Bank` from external subsystems (rpc, replay&banking). So, combined with `Bank::parent` fields, all of ancestors should be alive then none of them should be purged/pruned.

Haha yup I'm pretty sure I went through this exact line of reasoning when fixing up the scan the scan: https://github.com/solana-labs/solana/blob/master/runtime/src/accounts_index.rs#L415-L419

There's also a test for this here: https://github.com/solana-labs/solana/blob/master/runtime/src/bank.rs#L11533 which constantly scans unrooted slots that are being removed from BankForks by `squash()`

Reply:
> good question. From this discussion, I think we can revert to your original change and even simplify more.
> 
> In conclusion, I think we can promote the purging race-condition detection a hard error (https://github.com/solana-labs/solana/pull/15454/files?diff=split&w=1#r606630573).

Could you point me to the commit with this change, I don't think I see it yet.

Reply:
> These nice links aren't working for me cry

(github review functionality sucks. Usually, you can view what I'm referring to by opening a new browser window from the link if just clicking didn't work...)

Well, I waited for your go to promote this clause to `panic!()`, here's the freshly written commit: e02974b27a7afcafdfa34b54afcd69c072dc17c5

Reply:
^^ @carllin (I'm back working on this pr!)

Reply:
And finally I reverted to your original code: ecdf6ae1bc981a1884135482394ee818097f8ae0